File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    d = {chr(k):k-64 for k in range(65,91)}
    w = ''
    c = w
    num = 27
    for m in msg:
        c  = w + m
        if c in d:
            w = c
        else:
            answer.append(d[w])
            d[c] = num
            num += 1
            w = m
    if w in d:
        answer.append(d[w])
    return answer

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

vector<int> solution(string msg) {
    vector<int> answer;
    
    unordered_map<string,int> dict;
    for(int i=1;i<27;i++) {
        dict[string(1, 'A' + (i - 1))] = i;
    }
    
    int idx = 27;
    string w = "";
    w += msg[0];
    for(int i=1;i<msg.length();i++) {
        if (dict.count(w + msg[i])) {
            w += msg[i];
            continue;
        } else {
            answer.push_back(dict[w]);
            dict[w + msg[i]] = idx;
            w = msg[i];
            idx ++;
        }
    }
    answer.push_back(dict[w]);
    
    return answer;
}